### ISSUE TITLE
Succeeded given shorter input than matching string 

### ISSUE THREAD

**Issue by SunghoLee** [hbr/fmlib#1]
Unfortunately, a string parser seems to succeed in parsing a string value which is just a prefix of a parsing target string.

Here is a example for that.

```
let open Fmlib_parse.Character.Make (Unit) (String) (Unit) in
let p = Parser.run_on_string {|ab|} (make () (string {|abc|})) in
let remaining_tokens, is_ended = Parser.lookaheads p in
...
```

When executing the code, ``remaining_tokens`` is an empty array and ``is_ended`` is ``true``. Is this an intentional thing or a bug?
If this is intentional, how can I avoid this misbehavior?

**Comment by SunghoLee** [comment]
Oops, the example is not correct. In the above example, ``is_ended`` is ``true`` but ``Parser.has_succeeded p`` returns ``false``. 
However, there is still a wired point. Since ``string {|abc|}`` fails in parsing, ``backtrack (string {|abc|}) ""`` would push back all consumed tokens. But, when I combine it with ``string {|ab|}`` using ``</>``, the combined parser just consume ``a`` and ``b`` but fails in parsing. Here is a revised example:

```
let open Fmlib_parse.Character.Make (Unit) (String) (Unit) in
let p = Parser.run_on_string {|ab|} (make () ((backtrack (string {|abc|}) "") </> (string {|ab|})) in
let remaining_tokens, is_ended = Parser.lookaheads p in
let has_succeeded = Parser.has_succeeded p in
...
```

In this example, ``remaining_tokens`` is an empty array, but both ``is_ended`` and ``has_succeeded`` is ``false``. Is there some points that I misunderstand?

**Comment by hbr** [comment]
Thanks for reporting the issue. Let's analyze both examples.

1. The parser expects the three token 'a', 'b' and 'c' and then the end token. You have given the parser the two tokens 'a' and 'b' followed by the end token. The parser has consumed both character tokens and is expecting the character token 'c'. However it encounters the end token and enters into an error state. Therefore the parser has not succeeded. It has ended, because it cannot continue, but not succeeded. Remark: Your `is_ended` is not the information that parsing has ended. It just says that the end token has not yet been consumed.

    There are no more explicit lookahead tokens, therefore `remaining_tokens` is empty. However the end token has not been consumed. Therefore it is still on the lookahead.

2. The parser expects the three token 'a', 'b' and 'c' and then the end token of the token 'a', 'b' and then the end token. The first parser fails, because it encounters the end token instead of the expected token 'c'. Since it is a backtracking combinator, it shifts the token 'a', 'b' and the end token back to the input stream are parses the second combinator. The second combinator should succeed since it expects the token 'a' and 'b' followed by the end token. Therefore the behavior you observed is incorrect. I have reproduced your second example and found out that the parser has consumed the token 'a' and 'b' and is expecting the end token. Therefore I suspect some error in the parser in the combination of backtracking and the handling of the end token. The basic backtracking algorithm is correct, since it has been tested on many examples. The problem here seems to be that the backtracking is triggered by an unexpected end token.

As soon as I have a fix I come back to you.

Anyhow thanks for reporting the issue and giving a simple example with precise information.

Regards
Helmut

**Comment by SunghoLee** [comment]
Thank you for the fast response! This tool has really nice functionalities and APIs convenient for me to write a parser about complex syntax. 

Is there a workaround to parse both "abc" and "ab" in the current implementation?

**Comment by hbr** [comment]
I don't know of an easy workaround. But I have already identified the bug and the fix will come very soon.

**Comment by hbr** [comment]
There is a fix in the repo tagged with `0.3.1`. The official release usually takes some time. But you can use `opam pin ...` to update your installation with the fix.

**Comment by SunghoLee** [comment]
Thanks for the fast fix! But, I cannot get the ``0.3.1`` version from opam using ``opam pin ...``.
When I tried to ``opam pin fmlib_parse 0.3.1`` after ``opam update``, I got the follow error message:
```
[ERROR] Package fmlib_parse has no known version 0.3.1 in the repositories
```

Should I wait for the official release or is there another way to get it?

**Comment by hbr** [comment]
The version 0.3.1 has been released and is available in opam. You can get the new version with `opam upgrade fmlib_parse`. However before you upgrade you have to remove the pinned version with a command like `opam pin remove ...` (I don't know exactly the arguments).

The released version is identical to the pinned version.
